### DISCUSSION TITLE
Get embedded representation of each amino acid type

### DISCUSSION THREAD

**Discussion by hw-protein** [facebookresearch/esm#115] [Q&A] [answered]
First of all, I read your papers about the protein language model (ESM series). It's fascinating!

 I hope to use the amino acid type description from your pre-trained model. From your paper, "Biological structure and function emerge from scaling unsupervised learning to 250 million protein sequences" where you presented ESM-1b, you show the similarity between amino acid types with Figure 1. 
 
<img width="486" alt="스크린샷 2021-08-12 오전 11 07 28" src="https://user-images.githubusercontent.com/88811017/129128475-d265ea8c-dc80-47e1-8c82-75106ccb853b.png">

 I want to ask that "How did you extract the embedding representation for each amino acid type?". In your paper, you said "To investigate if the network has learned to encode physico- chemical properties in representations, we project the weight matrix of the final embedding layer of the network into two dimensions with t-SNE (Maaten & Hinton, 2008)." I do not have a deep knowledge of the deep learning field, so I do not understand it accurately.
 I hope to use the embedded representation of amino acids for other protein-related tasks. How can I get the vector(descriptor) that represents each amino acid type? 
 And can I extract same thing from ESM-1v ?

Thanks to read!

**Comment by tomsercu** [comment, marked as answer]
Hi! You can access the parameter tensor via `model.embed_tokens.weight`; in esm-1b and esm-1v the output and input embedding are shared. 
However, please beware that this is a relatively meaningless embedding for the AA in isolation! The real power of the model comes from processing the amino acid within the contact of the full protein sequence.
Good luck

**Reply by EBGU** [reply]
I am currently working on replicating your experiments. I extracted weight from ['model']['encoder.sentence_encoder.embed_tokens.weight'] of esm1v_t33_650M_UR90S_1.pt, and get the tok2idx as 
{'<cls>': 0, '<pad>': 1, '<eos>': 2, '<unk>': 3, 'L': 4, 'A': 5, 'G': 6, 'V': 7, 'S': 8, 'E': 9, 'R': 10, 'T': 11, 'I': 12, 'D': 13, 'P': 14, 'K': 15, 'Q': 16, 'N': 17, 'F': 18, 'Y': 19, 'M': 20, 'H': 21, 'W': 22, 'C': 23, 'X': 24, 'B': 25, 'U': 26, 'Z': 27, 'O': 28, '.': 29, '-': 30, '<null_1>': 31, '<mask>': 32}. 
However, when visualizing the embeddings[4:24] with TSNE from scikit-learn, I failed to get any meaningful plots. I thought I should have done something wrong, so could you kindly help me with it? Maybe provide your visualization script? Thank you very much!

**Reply by ZhangjLiang** [reply]
I also try to extract the embedding representation for each amino acid type but I don't know how to start and where to go from, have you made any progress lately please?
